File: PL_HW#1/PL_HW#01_20211397.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAX_TOKEN_LEN 100
#define MAX_TOKENS 100
#define MAX_LINE_LEN 1000

/* Token Code*/
#define INT_LIT 10
#define IDENT 11
#define ASSIGN_OP 20
#define ADD_OP 21
#define SUB_OP 22
#define MULT_OP 23
#define LEFT_PAREN 31
#define RIGHT_PAREN 32
#define SEMICOLON 33
#define END 34
#define PRINT 41

/*Token Type*/
#define LETTER 1
#define DIGIT 2
#define UNKNOWN 3

/* Token */
typedef struct _Token{
    int tokenCode;
    char lexeme[MAX_TOKEN_LEN];
}Token;

/* Global declararions */
char buffer[MAX_LINE_LEN];      //한줄씩 입력받기 버퍼
int tokenType;                  //토큰 타입 : 만약 토큰타입이 달라지면 Syntax Error ex)ab3
int tokenIndex;                 //현재 토큰 인덱스
int lexIndex;                   //lexeme 인덱스 
Token token[MAX_TOKENS];        //token 저장 전역 배열열
map <string, int> symbolTable;  //변수 저장용 
int currentTokenIndex;          //parser에서 토큰 추적 ind
int printed;                    //프린트호출여부(개행관리리)
bool errorFlag = false;         //에러 여부 확인 전역 변수
vector<int> printBuffer;        //프린트 버퍼


/* Function declarations */
int lookup(char);
int lexer(char*);
int findTokenType(char);

void program(void);
int statement(void);
int expr(void);
int term(void);
int factor(void);
int number(void);
string var(void);
bool digit(char);
bool alphabet(char);

Token currentToken();
void nextToken(void);
bool match(int);
int expect(int);

//=====================================================================================================
// lexer

/* -----------------------------------------------------------------------------------
 * 설명 : char을 입력으로 받아 해당 symbol이 정의되었는지 확인 하는 함수
 * 매계 : 확인하고자 하는 char c
 * 반환 : 성공 = 해당 symbol의 Token code , 실패 = -1
 * -----------------------------------------------------------------------------------
 */
int lookup(char c){
    switch(c){
        case '=':
            return ASSIGN_OP;
        case '+':
            return ADD_OP;
        case '-':
            return SUB_OP;
        case '*':
            return MULT_OP;
        case '(':
            return LEFT_PAREN;
        case ')':
            return RIGHT_PAREN;
        case ';':
            return SEMICOLON;
        default:
            return -1;
    }

}

/* -----------------------------------------------------------------------------------
 * 설명 : 토큰타입 구분 함수
 * 매계 : char c
 * 반환 : Token Type
 * -----------------------------------------------------------------------------------
 */
int findTokenType(char c){
    if(alphabet(c)){
        return LETTER;
    }
    else if(digit(c)){
        return DIGIT;
    }
    else{
        return UNKNOWN;
    }
}

/* -----------------------------------------------------------------------------------
 * 설명 : line을 입력으로 받아 tokenize, Token배열에 저장
 * 매계 : 입력받은 line
 * 반환 : 성공 = 0 , 실패 = -1
 * -----------------------------------------------------------------------------------
 */
int lexer(char* input) {
    int len = strlen(input);
    tokenIndex = 0;

    for (int i = 0; i < len;) {
        //공백제거
        while (i < len && isspace(input[i])) i++;
        if (i >= len) break;

        int currentType = findTokenType(input[i]);

        // UNKNOWN → symbol 처리
        if (currentType == UNKNOWN) {
            int code = lookup(input[i]);

            // 정의되지 않은 기호
            if (code == -1) {
                printf("Syntax Error!\n");
                return -1;
            }

            // 다음 글자도 공백 없이 붙으면 에러 (예: ++, =+ 등)
            if (i + 1 < len && !isspace(input[i + 1])) {
                printf("Syntax Error!\n");
                return -1;
            }

            // 단독 기호로 처리
            Token new_token;
            new_token.tokenCode = code;
            new_token.lexeme[0] = input[i++];
            new_token.lexeme[1] = '\0';
            token[tokenIndex++] = new_token;
            continue;
        }

        // 식별자 또는 숫자 처리
        Token new_token;
        lexIndex = 0;

        while (i < len && !isspace(input[i])) {
            int thisType = findTokenType(input[i]);
            if (thisType != currentType) {
                printf("Syntax Error!\n");
                return -1;
            }
            new_token.lexeme[lexIndex++] = input[i++];
        }

        new_token.lexeme[lexIndex] = '\0';

        // 토큰코드 매핑
        if (currentType == LETTER) {
            if (strcmp(new_token.lexeme, "print") == 0)
                new_token.tokenCode = PRINT;
            else
                new_token.tokenCode = IDENT;  // 일반 식별자도 IDENT 
        } else if (currentType == DIGIT) {
            new_token.tokenCode = INT_LIT;
        } else {
            printf("Syntax Error!\n");
            return -1;
        }

        token[tokenIndex++] = new_token;
    }
    // token이 끝남을 표시시
    token[tokenIndex].tokenCode = END;
    token[tokenIndex].lexeme[0] = '\0';

    return 0;
}

//=====================================================================================================
//parser

/*-----------------------------------------------------------------------------------
 * 설명 : 현재 토큰을 반환함
 * 반환 : 현재 토큰
 * -----------------------------------------------------------------------------------
 */
Token currentToken(void){
    return token[currentTokenIndex];
}

/*-----------------------------------------------------------------------------------
 * 설명 : tokenIndex 증가
 * -----------------------------------------------------------------------------------
 */
void nextToken(void){
    currentTokenIndex++;
}

/*-----------------------------------------------------------------------------------
 * 설명 : expected 값과 token Code를 비교
 *        ADD_OP 처럼 있어도 되도 없어도 되는경우에 사용하기 위해 에러메시지를 출력하는 expect와 분리
 * 반환 : 성공 : true, 실패 : false
 * -----------------------------------------------------------------------------------
 */
bool match(int expected){
    if(currentToken().tokenCode == expected){
        nextToken();
        return true;
    }
    return false;
}

/*-----------------------------------------------------------------------------------
 * 설명 : expected 값과 token Code를 비교 후 false 시 에러 처리
 * 반환 : 성공 : 0, 실패 : -1
 * -----------------------------------------------------------------------------------
 */
int expect(int expected){
    if(!match(expected)){
        errorFlag = true;
        return -1;
    }
    return 0;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <program> -> {<statement>}
 * -----------------------------------------------------------------------------------
 */
void program(void){
    symbolTable.clear();
    errorFlag = false;
    printed = 0;
    // 입력인 END가 아니고 에러가 없다면 statement()호출
    while(currentToken().tokenCode != END && !errorFlag){
        statement();
    }
    if(errorFlag){
        printf("Syntax Error!\n");
    }
    else if (printed) {
        if (!errorFlag) {
            for (int val : printBuffer) {
                printf("%d ", val);
            }
            putchar('\n');
        } 
        printBuffer.clear(); // 다음 줄 입력 대비
    }
}

/* -----------------------------------------------------------------------------------
 * 설명 : <statement> -> <var> = <expr> ; | print <var> ;
 * 반환 : 성공 :0, 실패 : -1
 * 전역변수 errorFlag로 에러여부 판단으로 수정함
 * -----------------------------------------------------------------------------------
 */
int statement(void){

    //print
    if(currentToken().tokenCode == PRINT){
        nextToken();
        //print 다음 <var>이 오지 않으면 에러
        if(currentToken().tokenCode != IDENT){
            errorFlag = true;
            return -1;
        }
            
        string varName = var();
        //print가 ';'로 끝나지 않을 시 에러
        if(expect(SEMICOLON)==-1) return -1;
        printed = 1;
        printBuffer.push_back(symbolTable[varName]);
    }
    else if(currentToken().tokenCode == IDENT){
        string varName = var(); 
        // '='기호 처리
        if(expect(ASSIGN_OP) == -1) return -1;
        
        //<expr> 처리 
        int res_expr = expr();
        if (errorFlag) return -1;

        // ';'기호 처리
        if(expect(SEMICOLON) == -1) return -1;

        //변수에 값 할당
        symbolTable[varName] = res_expr;
    }
    else{
        errorFlag = true;
        nextToken();
        return -1;
    }
    return 0;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <expr> -> <term> {+ <term> | * <term>}
 * 반환 : expr 값
 * -----------------------------------------------------------------------------------
 */
int expr(void){
    int res_term = term();
    if (errorFlag) return -1;

    //'+' or '*'기호 처리 
    while(!errorFlag && (currentToken().tokenCode == ADD_OP || currentToken().tokenCode == MULT_OP)){
        int op = currentToken().tokenCode;
        nextToken();
        int val = term();
        if(errorFlag) return -1;
        if(op == ADD_OP) res_term += val;
        else if(op == MULT_OP) res_term *= val;        
    }
    return res_term;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <term> -> <factor> {- <factor>}
 * 반환 : term값 
 * -----------------------------------------------------------------------------------
 */
int term(void){
    int res_factor = factor();
    if(errorFlag) return -1;

    // '-'기호 처리
    while(!errorFlag && (currentToken().tokenCode == SUB_OP)){
        nextToken();
        int val = factor();
        if(errorFlag) return -1;

        res_factor -= val;
    }
    return res_factor;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <factor> -> [ - ] ( <number> | <var> | '('<expr>')')
 * 반환 : factor값
 * -----------------------------------------------------------------------------------
 */
int factor(void){
    int sign = 1;

    if(currentToken().tokenCode == SUB_OP){
        sign = -1;
        nextToken();

        //연속된 - 에러처리
        if (currentToken().tokenCode == SUB_OP){
            errorFlag = true;
            nextToken();
            return -1;
        }
    }
    
    // number
    if(currentToken().tokenCode == INT_LIT){
        int val = number();
        return sign*val;
    }
    //var
    else if(currentToken().tokenCode == IDENT){
        string varName = var();
        return sign*symbolTable[varName];
    }
    //괄호
    else if(currentToken().tokenCode == LEFT_PAREN){
        nextToken();
        int val = expr();
        if(expect(RIGHT_PAREN) == -1){
            nextToken();
            return -1;
        }
        return sign * val;
    }
    else{
        errorFlag = true;
        nextToken();
        return -1;
    }
}

/* -----------------------------------------------------------------------------------
 * 설명 : <number> -> <digit> {<digit>}
 * -----------------------------------------------------------------------------------
 */
int number(void){
    int val = atoi(currentToken().lexeme);
    nextToken();
    return val;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <var> -> <alphabet>{<alphabet>}
 * -----------------------------------------------------------------------------------
 */
string var(void){
    if (currentToken().tokenCode != IDENT){
        errorFlag = true;
        return "";
    }
    string varName = currentToken().lexeme;
    nextToken();
    return varName;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <statement> -> a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s | t | u | v | w | x | y | z
 * -----------------------------------------------------------------------------------
 */
bool alphabet(char c){
    char al[26] = {'a','b','c','d','e','f','g','h',
                 'i','j','k','l','m','n','o','p',
                 'q','r','s','t','u','v','w','x','y','z'};
    for(int i = 0 ; i < 26 ; i++){
        if (c == al[i]){
            return true;
        }
    }
    return false;
}

/* -----------------------------------------------------------------------------------
 * 설명 : <digit> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 
 * -----------------------------------------------------------------------------------
 */
bool digit(char c){
    if(isdigit(c)){
        return true;
    }
    return false;
}

int main(void){
    while(true){
        //함수 시작 전 global var 초기화
        tokenIndex = 0;
        tokenType = 0;
        lexIndex = 0;
        currentTokenIndex = 0;

        cout << ">> ";
        cin.getline(buffer, MAX_LINE_LEN);

        if (strlen(buffer) == 0) break;
        if (lexer(buffer) != -1){
            program();
            errorFlag = false;
        }
    }
}

File: PL_HW#1/PL_HW#01_20211397.py
# Token Code
token_codes = {
    'INT_LIT': 10,
    'IDENT': 11,
    'ASSIGN_OP': 20,
    'ADD_OP': 21,
    'SUB_OP': 22,
    'MULT_OP': 23,
    'LEFT_PAREN': 31,
    'RIGHT_PAREN': 32,
    'SEMICOLON': 33,
    'END': 34,
    'PRINT': 41
}

# Token class
class Token:
    def __init__(self, tokenCode, lexeme):
        self.tokenCode = tokenCode
        self.lexeme = lexeme


#Global declararion
token = []             #한줄씩 입력받기 버퍼
tokenIndex = 0         #현재 토큰 인덱스
currentTokenIndex = 0  #parser에서 토큰 추적 index
symbol_table = {}      #변수 저장용 심볼테이블
printed = False        #프린트호출여부(개행관리)
errorFlag = False      #에러여부확인
printBuffer = []       #프린트 버퍼퍼

'''
-----------------------------------------------------------------------------------
설명 : 토큰타입 구분 함수
매계 : char c
반환 : Token Type
-----------------------------------------------------------------------------------
'''
def findTokenType(c):
    if alphabet(c): 
        return 'LETTER'
    if digit(c): 
        return 'DIGIT'
    return 'UNKNOWN'

'''
====================================================================================
lexer
'''

'''
-----------------------------------------------------------------------------------
설명 : char을 입력으로 받아 해당 symbol이 정의되었는지 확인 하는 함수
매계 : 확인하고자 하는 char c
반환 : 성공 = 해당 symbol의 Token code , 실패 = -1
-----------------------------------------------------------------------------------
'''
def lookup(c):
    return {
        '=': token_codes['ASSIGN_OP'],
        '+': token_codes['ADD_OP'],
        '-': token_codes['SUB_OP'],
        '*': token_codes['MULT_OP'],
        '(': token_codes['LEFT_PAREN'],
        ')': token_codes['RIGHT_PAREN'],
        ';': token_codes['SEMICOLON']
    }.get(c, -1)


'''
-----------------------------------------------------------------------------------
설명 : line을 입력으로 받아 tokenize, Token배열에 저장
매계 : 입력받은 line
반환 : 성공 = 0 , 실패 = -1
-----------------------------------------------------------------------------------
'''
def lexer(input_line):
    global token, tokenIndex
    token.clear()   #입력받을 버퍼(token) 초기화
    tokenIndex = 0
    i = 0
    length = len(input_line)

    while i < length:
        #공백제거
        while i < length and input_line[i].isspace():
            i += 1
        if i >= length: 
            break

        c = input_line[i]
        currentType = findTokenType(c)

        # UNKNOWN -> symbol 처리
        if currentType == 'UNKNOWN':
            code = lookup(c)

            #정의되지 않은 기호
            if code == -1:
                print("Syntax Error!")
                return -1

            #다음글자가 공백없이 붙으면 에러(++, =+ 등)
            if i + 1 < length and not input_line[i+1].isspace():
                print("Syntax Error!")
                return -1
            token.append(Token(code, c))
            i += 1
        else:
            lexeme = ''
            while i < length and findTokenType(input_line[i]) == currentType:
                lexeme += input_line[i]
                i += 1

        #토큰코드 매핑 
            #LETTER일 때 print 따로 처리
            if currentType == 'LETTER':
                code = token_codes['PRINT'] if lexeme == 'print' else token_codes['IDENT']
            #DIGIT는 INT_LIT 토큰
            elif currentType == 'DIGIT':
                code = token_codes['INT_LIT']
            else:
                print("Syntax Error!")
                return -1

            token.append(Token(code, lexeme))

    token.append(Token(token_codes['END'], ''))
    return 0


'''
====================================================================================
parser
'''

'''
-----------------------------------------------------------------------------------
설명 : 현재 토큰을 반환함
반환 : 현재 토큰
----------------------------------------------------------------------------------
'''
def currentToken():
    return token[currentTokenIndex]

'''
-----------------------------------------------------------------------------------
설명 : tokenIndex 증가
-----------------------------------------------------------------------------------
'''
def nextToken():
    global currentTokenIndex
    currentTokenIndex += 1
'''
-----------------------------------------------------------------------------------
설명 : expected 값과 token Code를 비교
       ADD_OP 처럼 있어도 되도 없어도 되는경우에 사용하기 위해 에러메시지를 출력하는 expect와 분리
반환 : 성공 : true, 실패 : false
-----------------------------------------------------------------------------------
'''
def match(expected):
    if currentToken().tokenCode == expected:
        nextToken()
        return True
    return False
    

'''
-----------------------------------------------------------------------------------
설명 : expected 값과 token Code를 비교 후 false 시 에러 처리
반환 : 성공 : 0, 실패 : -1
-----------------------------------------------------------------------------------
'''
def expect(expected):
    global errorFlag
    if not match(expected):
        errorFlag = True
        return -1
    return 0

# Grammar rules
'''
-----------------------------------------------------------------------------------
설명 : <program> -> {<statement>}
-----------------------------------------------------------------------------------
'''
def program():
    global errorFlag, printed, printBuffer, symbol_table
    symbol_table.clear()
    errorFlag = False
    printed = False
    
    #입력인 END가 아니고 에러가 없다면 statement()호출
    while currentTokenIndex < len(token) and currentToken().tokenCode != token_codes['END'] and not errorFlag:
        statement()
    if errorFlag:
        print("Syntax Error!")
    elif printed:
        if not errorFlag:
            #버퍼에 있는 값 합쳐서 출력
            print(' '.join(map(str, printBuffer)))
    printBuffer.clear()

'''
-----------------------------------------------------------------------------------
설명 : <statement> -> <var> = <expr> ; | print <var> ;
반환 : 성공 :0, 실패 : -1
전역변수 errorFlag로 에러여부 판단으로 수정함
-----------------------------------------------------------------------------------
'''
def statement():
    global printed, errorFlag

    #print
    if currentToken().tokenCode == token_codes['PRINT']:
        nextToken()
        #print 뒤에 <var>이 아닌경우 error
        if currentToken().tokenCode != token_codes['IDENT']:
            errorFlag = True
            return -1
        varName = var()
        #print가 ;로 끝나지 않으면 error
        if expect(token_codes['SEMICOLON']) == -1: 
            return -1
        printed = True
        printBuffer.append(symbol_table.get(varName, 0))
        
    # <var> = <expr> ;
    elif currentToken().tokenCode == token_codes['IDENT']:
        varName = var()
        # '='기호 처리
        if expect(token_codes['ASSIGN_OP']) == -1: 
            return -1
        #<expr> 호출
        val = expr()
        if errorFlag: 
            return -1

        # ; 처리
        if expect(token_codes['SEMICOLON']) == -1: 
            return -1

        # 심볼테이블에 변수 값 할당
        symbol_table[varName] = val
    else:
        errorFlag = True
        nextToken()
        return -1
    return 0

'''
-----------------------------------------------------------------------------------
설명 : <expr> -> <term> {+ <term> | * <term>}
반환 : expr 값 실패시 -1
-----------------------------------------------------------------------------------
'''
def expr():
    res_term = term()
    if errorFlag:
        return -1
    while not errorFlag and currentToken().tokenCode in [token_codes['ADD_OP'], token_codes['MULT_OP']]:
        op = currentToken().tokenCode
        nextToken()
        val = term()
        if errorFlag: 
            return -1
        #ADD_OP이면 더하기
        if op == token_codes['ADD_OP']:
            res_term += val
        #MULT_OP이면 곱하기
        elif op == token_codes['MULT_OP']:
            res_term *= val
    return res_term

'''
-----------------------------------------------------------------------------------
설명 : <term> -> <factor> {- <factor>}
반환 : term값 
-----------------------------------------------------------------------------------
'''
def term():
    res_factor = factor()
    if errorFlag:
        return -1

    # - 기호 처리
    while not errorFlag and currentToken().tokenCode == token_codes['SUB_OP']:
        nextToken()
        val = factor()
        if errorFlag:
            return -1
        res_factor -= val
    return res_factor

'''
-----------------------------------------------------------------------------------
설명 : <factor> -> [ - ] ( <number> | <var> | '('<expr>')')
       [-]처리를 위해 sign 변수를 선언해 곱해줌
반환 : factor값 실패시 -1
-----------------------------------------------------------------------------------
'''
def factor():
    global errorFlag
    sign = 1
    if currentToken().tokenCode == token_codes['SUB_OP']:
        sign = -1
        nextToken()

        # 연속된 - 에러 처리 (ex) - - )
        if currentToken().tokenCode == token_codes['SUB_OP']:
            errorFlag = True
            nextToken()
            return -1

    # number처리
    if currentToken().tokenCode == token_codes['INT_LIT']:
        return sign * number()

    # var 처리
    elif currentToken().tokenCode == token_codes['IDENT']:
        return sign * symbol_table.get(var(), 0)

    # 괄호 처리
    elif currentToken().tokenCode == token_codes['LEFT_PAREN']:
        nextToken()
        val = expr()
        if expect(token_codes['RIGHT_PAREN']) == -1:
            nextToken()
            return -1
        return sign * val
    else:
        errorFlag = True
        nextToken()
        return -1

'''
-----------------------------------------------------------------------------------
설명 : <number> -> <digit> {<digit>}
-----------------------------------------------------------------------------------
'''
def number():
    val = int(currentToken().lexeme)
    nextToken()
    return val

'''
-----------------------------------------------------------------------------------
설명 : <var> -> <alphabet>{<alphabet>}
-----------------------------------------------------------------------------------
'''
def var():
    global errorFlag
    if currentToken().tokenCode != token_codes['IDENT']:
        errorFlag = True
        return ""
    varName = currentToken().lexeme
    nextToken()
    return varName

'''
-----------------------------------------------------------------------------------
설명 : <statement> -> a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s | t | u | v | w | x | y | z
-----------------------------------------------------------------------------------
'''
def alphabet(c):
    al = ('a','b','c','d','e','f','g','h',
        'i','j','k','l','m','n','o','p',
        'q','r','s','t','u','v','w','x','y','z')
    if c in al:
        return True
    return False

'''
-----------------------------------------------------------------------------------
설명 : <digit> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 
-----------------------------------------------------------------------------------
'''
def digit(c):
    digits = ('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')
    if c in digits:
        return True
    return False

# Main loop
while True:
    try:
        line = input(">> ")
        if not line.strip():
            break
        currentTokenIndex = 0
        if lexer(line) != -1:
            program()
    except EOFError:
        break
